sapicore/model/__init__.py

```diff
@@ -1,120 +1,169 @@
-"""Model
-========
+"""Fit, test, and use neuromorphic models.
 
-A :class:`SapicoreModel` is a higher level abstraction of a
-:class:`~sapicore.network.SapicoreNetwork` that can also contain
-sub-:attr:`~SapicoreModel.networks` that are connected to each other via
-neurons at the input/output boundaries of the sub-networks.
+Models are networks endowed with the methods :meth:`fit`, :meth:`predict`, :meth:`similarity`, :meth:`load`,
+and :meth:`save`. This class provides an ML-focused API for the training and usage of sapicore
+:class:`~engine.network.Network` output for practical purposes.
 
-A :class:`SapicoreModel` additionally contains a learning element that can be
-added and applied to the model.
 """
-from typing import Dict, List, Optional, Tuple
-from sapicore.network import SapicoreNetwork
-from sapicore.neuron import SapicoreNeuron
-from sapicore.learning import SapicoreLearning
+from typing import Callable
 
-__all__ = ('SapicoreModel', )
+import dill
```

Comment:
I'm suuuper weary of using something like dill or pickle to dump our objects. This brings a ton of issues and I think we're better off without it. The issues begin with data consistency, where you have to make sure the versions of everything you are loading and its dependencies match that of the dumped environment and end with security issues.

If we are using something to pickle our state, which is impossible to do securely btw, we should at least be very specific in what can get pickled. But that is an undertaking. It's better to rely on the yaml flow instead IMO.

Reply:
The convenience is almost too tempting, but I agree it's worth spending the extra time to bypass pickle/dill. I had assumed that PyTorch's load/save (which use pickle/dill) somehow deal with or mitigate potential issues, but after reading a few discussions, it does seem like a suboptimal decision on their part.

Reply:
Do we need it rn? Can we leave it out and figure it out later? Because theoretically we should be able to generate yaml with all the details!? Or am I wrong?

Reply:
We don't strictly need to serialize the full object, but we do need a way of reloading or reconstructing trained models from configurables and tensor buffers (weights etc.) during testing. The only potential snag (that dill could handle out of the box) is the `self.equation` lambda expression in Neuron.

Reply:
Yeah, that's where things start. Rn it may only be `self.equation` but later it could be other objects or even object from external dependencies that gets included. That's why I think it's better to not have anything like pickle at all and instead rely on config properties in yaml and potentially arrays stored in hdf5 (or yaml). In this approach, you wouldn't save `self.equation` because the equation is set based on a config property.

So, if you need to share models, they'd be linked to specific commits, including a poetry lock file. This will ensure that config can completely reconstruct the objects without having to rely on pickling actual objects. And the onus will be on model developers to make sure you can reconstruct a model purely from config and data arrays.

Reply:
Right--instantiating a neuron will take care of self.equation, and that can be done with just the configurables and voltage/weight/etc. buffers. What do you think about adding generic save/load methods to `Component` that would handle the dumping/reconstructing of configurables and loggables to/from HDF? Seems to be the most efficient solution. Then, users may override it if they want to save to YAMLs or some other format.

Agree with the rest--this circles back to the project directory format discussion. "Result" commits might include  multiple timestamped run directories containing snapshots of models trained on different sets/folds (plus original config files, docs, and other deliverables like analyses and plots). I suppose we'll converge once we get going.

Reply:
@matham Are you open to leaving torch.load/save in until 0.3.1? The NeurIPS deadline is approaching and we need a quick and dirty way to save and load trained models. I'll open an issue to make sure we don't forget it needs to be fixed.

If not, I'll have those methods pass (rather than `raise NotImplementedError`) to avoid breaking some tests.

Reply:
Sure, we can leave it in for now.